libraries/AP_MSP/msp.cpp

```diff
@@ -0,0 +1,296 @@
+            if (c == '$') {
+                msp->msp_version = MSP_V1;
+                msp->c_state = MSP_HEADER_START;
+            } else {
+                return false;
+            }
+            break;
+
+        case MSP_HEADER_START:  // Waiting for 'M' (MSPv1 / MSPv2_over_v1) or 'X' (MSPv2 native)
+            switch (c) {
+                case 'M':
+                    msp->c_state = MSP_HEADER_M;
+                    break;
+                case 'X':
+                    msp->c_state = MSP_HEADER_X;
+                    break;
+                default:
+                    msp->c_state = MSP_IDLE;
+                    break;
+            }
+            break;
+
+        case MSP_HEADER_M:      // Waiting for '<'
+            if (c == '<') {
+                msp->offset = 0;
+                msp->checksum1 = 0;
+                msp->checksum2 = 0;
+                msp->c_state = MSP_HEADER_V1;
+            } else {
+                msp->c_state = MSP_IDLE;
+            }
+            break;
+
+        case MSP_HEADER_X:
+            if (c == '<') {
+                msp->offset = 0;
+                msp->checksum2 = 0;
+                msp->msp_version = MSP_V2_NATIVE;
+                msp->c_state = MSP_HEADER_V2_NATIVE;
+            } else {
+                msp->c_state = MSP_IDLE;
+            }
+            break;
+
+        case MSP_HEADER_V1:     // Now receive v1 header (size/cmd), this is already checksummable
+            msp->in_buf[msp->offset++] = c;
+            msp->checksum1 ^= c;
+            if (msp->offset == sizeof(msp_header_v1_t)) {
+                msp_header_v1_t * hdr = (msp_header_v1_t *)&msp->in_buf[0];
+                // Check incoming buffer size limit
+                if (hdr->size > MSP_PORT_INBUF_SIZE) {
+                    msp->c_state = MSP_IDLE;
+                } else if (hdr->cmd == MSP_V2_FRAME_ID) {
+                    // MSPv1 payload must be big enough to hold V2 header + extra checksum
+                    if (hdr->size >= sizeof(msp_header_v2_t) + 1) {
+                        msp->msp_version = MSP_V2_OVER_V1;
+                        msp->c_state = MSP_HEADER_V2_OVER_V1;
+                    } else {
+                        msp->c_state = MSP_IDLE;
+                    }
+                } else {
+                    msp->data_size = hdr->size;
+                    msp->cmd_msp = hdr->cmd;
+                    msp->cmd_flags = 0;
+                    msp->offset = 0;                // re-use buffer
+                    msp->c_state = msp->data_size > 0 ? MSP_PAYLOAD_V1 : MSP_CHECKSUM_V1;    // If no payload - jump to checksum byte
+                }
+            }
+            break;
+
+        case MSP_PAYLOAD_V1:
+            msp->in_buf[msp->offset++] = c;
+            msp->checksum1 ^= c;
+            if (msp->offset == msp->data_size) {
+                msp->c_state = MSP_CHECKSUM_V1;
+            }
+            break;
+
+        case MSP_CHECKSUM_V1:
```

Comment:
Interestingly, `MSP_HEADER_X` means, "we have received an X header`, but `MSP_CHECKSUM_V1` means, "I want to receive a checksum byte now"

Reply:
The MSP decoder code has many of these because the code has been "poorly" maintained, new MSP messages and extension can come from betaflight, cleanflight, inav, etc... these is a known issue even among the code maintainers